udpsession.go

```diff
@@ -0,0 +1,659 @@
+	us.RUnlock()
+
+	us.Lock()
+	defer us.Unlock()
+	if us.isListener {
+		us.sessKeyToSessAddr = make(map[string]*udpSessionAddr)
+		us.udpAddrToSessAddr = make(map[string]*udpSessionAddr)
+		for _, conn := range us.tcpConns {
+			if conn != nil {
+				conn.Close()
+			}
+		}
+		us.tcpConns = make(map[string]*Conn)
+	} else {
+		us.peer = nil
+		if us.tcpConn != nil {
+			return us.tcpConn.Close()
+		}
+	}
+
+	return
+}
+
+func (us *UdpSession) sendCloseMsg(udpAddr *net.UDPAddr) (err error) {
+	b := newSessionData(pb.MsgType_CLOSE, nil)
+
+	us.RLock()
+	defer us.RUnlock()
+
+	if !us.isListener { // Dialer
+		_, err = us.tcpConn.Write(b)
+		return
+	}
+
+	if udpAddr != nil { // Send close message to this dialer address
+		if usa, ok := us.udpAddrToSessAddr[udpAddr.String()]; ok {
+			if tcpConn, ok := us.tcpConns[usa.String()]; ok {
+				_, err = tcpConn.Write(b)
+			}
+		}
+		return
+	}
+
+	for _, usa := range us.sessKeyToSessAddr { // Send close message to all dialers.
+		if tcpConn, ok := us.tcpConns[usa.String()]; ok {
+			_, err = tcpConn.Write(b)
+		}
+	}
+
+	return
+}
+
+// Handle connection closed by peer or getting closed message from peer.
+func (us *UdpSession) handleConnClosed(sessKey string) {
+	us.Lock()
+	defer us.Unlock()
+	if us.isListener {
+		if usa, ok := us.sessKeyToSessAddr[sessKey]; ok {
+			delete(us.sessKeyToSessAddr, sessKey)
+			delete(us.udpAddrToSessAddr, usa.udpAddr.String())
+			if tcpConn, ok := us.tcpConns[sessKey]; ok {
+				tcpConn.Close()
+				delete(us.tcpConns, sessKey)
+			}
+		}
+	} else {
+		us.peer = nil
+		us.udpConn.Close()
+		us.tcpConn.Close()
+	}
+}
+
+// Get UDPAddr by net.Addr
+func (us *UdpSession) getUdpAddr(addr net.Addr) (udpAddr *net.UDPAddr, err error) {
+	us.RLock()
+	defer us.RUnlock()
+
+	if !us.isListener {
+		udpAddr = us.peer.udpAddr
+		return
+	}
+	// for listener
+	if addr == nil {
+		return nil, ErrWrongAddr
+	} else {
+		usa, ok := addr.(*udpSessionAddr) // addr should be *udpSessionAddr
+		if !ok {
+			return nil, ErrWrongAddr
+		}
+
+		usa, ok = us.sessKeyToSessAddr[usa.String()]
+		if ok {
+			udpAddr = usa.udpAddr
+		} else {
+			return nil, ErrWrongAddr
+		}
+	}
+
+	return
+}
+
+func (us *UdpSession) sendMetaData(sessionID []byte) (err error) {
+	metadata := &pb.SessionMetadata{
+		Id:            sessionID,
+		SessionType:   pb.SessionType_UDP,
+		DialerNknAddr: us.ts.Address(),
+	}
+	buf, err := proto.Marshal(metadata)
+	if err != nil {
+		return err
+	}
+
+	b := newSessionData(pb.MsgType_SESSIONID, buf)
+
+	us.RLock()
+	_, _, err = us.udpConn.WriteMsgUDP(b, nil, nil)
+	us.RUnlock()
+
+	return err
+}
+
+// Handle session meta data.
+func (us *UdpSession) handleMetaData(b []byte, udpAddr *net.UDPAddr) error {
+	header := b[0]
+	if header != byte(pb.HeaderType_SESSION) {
+		return ErrWrongMsgFormat
+	}
+	sessData, err := parseSessionData(b)
+	if err != nil {
+		return err
+	}
+
+	switch sessData.MsgType {
+	case pb.MsgType_SESSIONID:
+		metadata := &pb.SessionMetadata{}

```

Comment:
Each conn should only handle metadata once. The rest should be ignored.
Also, dialer side should always ignore metadata.